lib/internal/test_runner/yaml_parser.js

```diff
@@ -0,0 +1,109 @@
+'use strict';
+const {
+  codes: {
+    ERR_TEST_FAILURE,
+  }
+} = require('internal/errors');
+const AssertionError = require('internal/assert/assertion_error');
+const {
+  ArrayPrototypeJoin,
+  Error,
+  Number,
+  NumberIsNaN,
+  ObjectAssign,
+  RegExpPrototypeExec,
+  StringPrototypeEndsWith,
+  StringPrototypeRepeat,
+  StringPrototypeSlice,
+  StringPrototypeStartsWith,
+  StringPrototypeSubstring,
+} = primordials;
+
+const kYamlKeyRegex = /^(\s+)?(\w+):(\s)+([>|][-+])?(.*)$/;
+const kStackDelimiter = '    at ';
+
+function reConstructError(parsedYaml) {
+  if (!('error' in parsedYaml)) {
+    return parsedYaml;
+  }
+  const isAssertionError = parsedYaml.code === 'ERR_ASSERTION' ||
+    'actual' in parsedYaml || 'expected' in parsedYaml || 'operator' in parsedYaml;
+  const isTestFailure = parsedYaml.code === 'ERR_TEST_FAILURE' || 'failureType' in parsedYaml;
+  const stack = parsedYaml.stack ? kStackDelimiter + ArrayPrototypeJoin(parsedYaml.stack, `\n${kStackDelimiter}`) : '';
+  let error;
+
+  if (isAssertionError) {
+    error = new AssertionError({
+      message: parsedYaml.error,
+      actual: parsedYaml.actual,
+      expected: parsedYaml.expected,
+      operator: parsedYaml.operator
+    });
+  }
+
+  if (isTestFailure) {
+    // eslint-disable-next-line no-restricted-syntax
+    const cause = error ?? ObjectAssign(new Error(parsedYaml.error), { code: parsedYaml.code });
+    cause.stack = stack;
+    error = new ERR_TEST_FAILURE(cause, parsedYaml.failureType);
+    error.stack = stack;
+  }
+
+  parsedYaml.error = error;
+  delete parsedYaml.stack;
+  delete parsedYaml.code;
+  delete parsedYaml.failureType;
+  delete parsedYaml.actual;
+  delete parsedYaml.expected;
+  delete parsedYaml.operator;
+
+  return parsedYaml;
+}
+
+function getYamlValue(value, key) {
+  if (StringPrototypeStartsWith(value, "'") && StringPrototypeEndsWith(value, "'")) {
+    return StringPrototypeSlice(value, 1, -1);
+  }
+  if (value === 'true') {
+    return true;
+  }
+  if (value === 'false') {
+    return false;
+  }
```

Comment:
doesn't yaml also support a bunch of other strings like on/off/etc?

Reply:
it does but we currently only emit a subset of YAML.
this parser is designed to handle the YAML generated by https://github.com/nodejs/node/blob/a996638e53c82a7c60589f88c6d7b517e9fd5505/lib/internal/test_runner/tap_stream.js#L145
not to be a full-blown YAML parser